File: personal-dev-projects/gravity-flow-engine/examples/physics-core-demo/physics-engine.js
/**
 * Gravity Flow Engine - Core Physics Implementation
 * 
 * This file implements the mathematical foundation of the physics engine
 * without any visualization components.
 */

class PhysicsEngine {
  constructor(nodes, links, options = {}) {
    // Deep clone nodes and add required physics properties
    this.nodes = nodes.map(node => ({
      ...node,
      x: node.x || Math.random() * 1000,
      y: node.y || Math.random() * 1000,
      vx: 0,
      vy: 0,
      fx: null,
      fy: null
    }));
    
    // Convert link references to actual node objects
    this.links = links.map(link => ({
      ...link,
      source: typeof link.source === 'string' ? this.findNodeById(link.source) : link.source,
      target: typeof link.target === 'string' ? this.findNodeById(link.target) : link.target
    }));
    
    // Physics parameters
    this.options = {
      alpha: 1.0,              // Current "temperature" of simulation
      alphaMin: 0.001,         // Minimum alpha before simulation stops
      alphaDecay: 0.0228,      // Alpha reduction per tick
      alphaTarget: 0,          // Target alpha value
      velocityDecay: 0.4,      // Node velocity decay factor (drag)
      ...options
    };
    
    this.forces = {};
    this.tickCount = 0;
  }
  
  findNodeById(id) {
    return this.nodes.find(n => n.id === id);
  }
  
  // Register a force function
  addForce(name, forceFn) {
    this.forces[name] = forceFn;
    return this;
  }
  
  // Core physics integration step using Velocity Verlet
  tick() {
    // Skip if simulation has cooled down
    if (this.options.alpha < this.options.alphaMin) return false;
    
    this.tickCount++;
    
    // Apply forces to calculate acceleration
    Object.values(this.forces).forEach(force => {
      force(this.options.alpha);
    });
    
    // Update positions using Velocity Verlet integration
    this.nodes.forEach(node => {
      if (node.fx !== null) {
        node.x = node.fx;
        node.vx = 0;
      } else {
        node.vx *= this.options.velocityDecay;
        node.x += node.vx;
      }
      
      if (node.fy !== null) {
        node.y = node.fy;
        node.vy = 0;
      } else {
        node.vy *= this.options.velocityDecay;
        node.y += node.vy;
      }
    });
    
    // Cool down simulation
    this.options.alpha += (this.options.alphaTarget - this.options.alpha) * this.options.alphaDecay;
    
    return true;
  }
  
  // Standard forces implementation
  
  // Repulsive force between nodes (like charges)
  createManyBodyForce(strength = -30) {
    return (alpha) => {
      for (let i = 0; i < this.nodes.length; i++) {
        const node1 = this.nodes[i];
        
        for (let j = i + 1; j < this.nodes.length; j++) {
          const node2 = this.nodes[j];
          
          // Calculate distance vector
          const dx = node2.x - node1.x;
          const dy = node2.y - node1.y;
          const distanceSquared = dx * dx + dy * dy;
          
          // Skip if nodes are at the same position
          if (distanceSquared === 0) continue;
          
          // Calculate repulsive force (inverse square law)
          const distance = Math.sqrt(distanceSquared);
          const force = strength * node1.physicalProperties.charge * node2.physicalProperties.charge / distanceSquared;
          
          // Apply force to both nodes (Newton's third law)
          const unitX = dx / distance;
          const unitY = dy / distance;
          
          node1.vx -= unitX * force * alpha;
          node1.vy -= unitY * force * alpha;
          node2.vx += unitX * force * alpha;
          node2.vy += unitY * force * alpha;
        }
      }
    };
  }
  
  // Spring force between linked nodes
  createLinkForce() {
    return (alpha) => {
      this.links.forEach(link => {
        const source = link.source;
        const target = link.target;
        
        // Calculate distance vector
        const dx = target.x - source.x;
        const dy = target.y - source.y;
        const distance = Math.sqrt(dx * dx + dy * dy);
        
        // Skip if nodes are at the same position
        if (distance === 0) return;
        
        // Calculate spring force (Hooke's law)
        const naturalLength = link.physicalProperties.length;
        const stiffness = link.physicalProperties.stiffness;
        const displacement = distance - naturalLength;
        const springForce = stiffness * displacement;
        
        // Apply force proportional to displacement
        const unitX = dx / distance;
        const unitY = dy / distance;
        
        const fx = springForce * unitX * alpha;
        const fy = springForce * unitY * alpha;
        
        source.vx += fx;
        source.vy += fy;
        target.vx -= fx;
        target.vy -= fy;
      });
    };
  }
  
  // Force that pushes nodes toward a center point
  createCenterForce(centerX, centerY, strength = 0.1) {
    return (alpha) => {
      this.nodes.forEach(node => {
        node.vx += (centerX - node.x) * strength * alpha;
        node.vy += (centerY - node.y) * strength * alpha;
      });
    };
  }
  
  // Collision detection force
  createCollisionForce() {
    return (alpha) => {
      for (let i = 0; i < this.nodes.length; i++) {
        const node1 = this.nodes[i];
        const radius1 = node1.physicalProperties.radius;
        
        for (let j = i + 1; j < this.nodes.length; j++) {
          const node2 = this.nodes[j];
          const radius2 = node2.physicalProperties.radius;
          
          // Calculate distance vector
          const dx = node2.x - node1.x;
          const dy = node2.y - node1.y;
          const distance = Math.sqrt(dx * dx + dy * dy);
          
          // Skip if nodes are far apart
          const minDistance = radius1 + radius2;
          if (distance >= minDistance) continue;
          
          // Calculate collision response
          const unitX = dx / distance;
          const unitY = dy / distance;
          
          // Move nodes apart
          const separation = minDistance - distance;
          const moveX = unitX * separation * 0.5;
          const moveY = unitY * separation * 0.5;
          
          node1.vx -= moveX * alpha;
          node1.vy -= moveY * alpha;
          node2.vx += moveX * alpha;
          node2.vy += moveY * alpha;
        }
      }
    };
  }
  
  // Initialize the physics engine with standard forces
  initializeStandardForces(width, height) {
    return this
      .addForce("charge", this.createManyBodyForce())
      .addForce("link", this.createLinkForce())
      .addForce("center", this.createCenterForce(width/2, height/2))
      .addForce("collision", this.createCollisionForce());
  }
  
  // Run simulation for a certain number of steps
  runSimulation(steps) {
    let i = 0;
    for (i = 0; i < steps; i++) {
      if (!this.tick()) break;
    }
    return {
      nodes: this.nodes,
      completedSteps: i,
      alpha: this.options.alpha
    };
  }
  
  // Get current state of the simulation
  getState() {
    return {
      nodes: this.nodes.map(n => ({
        id: n.id,
        x: n.x,
        y: n.y,
        vx: n.vx,
        vy: n.vy,
        category: n.category
      })),
      links: this.links.map(l => ({
        source: l.source.id,
        target: l.target.id,
        length: l.physicalProperties.length
      })),
      tickCount: this.tickCount,
      alpha: this.options.alpha
    };
  }
}

module.exports = PhysicsEngine; 

File: personal-dev-projects/gravity-flow-engine/examples/physics-core-demo/rust-impl/src/physics_engine.rs
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// Core types for the physics engine
type NodeId = String;
type ForceFunction = fn(&mut PhysicsEngine, f64);

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PhysicalProperties {
    pub mass: f64,
    pub charge: f64,
    pub friction: f64,
    pub radius: f64,
    pub fixed: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LinkPhysicalProperties {
    pub stiffness: f64,
    pub length: f64,
    pub elasticity: f64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Node {
    pub id: NodeId,
    pub label: String,
    pub category: String,
    pub physical_properties: PhysicalProperties,
    // Physics state
    pub x: f64,
    pub y: f64,
    pub vx: f64,
    pub vy: f64,
    pub fx: Option<f64>,
    pub fy: Option<f64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Link {
    pub source: NodeId,
    pub target: NodeId,
    pub physical_properties: LinkPhysicalProperties,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct NodeState {
    pub id: NodeId,
    pub x: f64,
    pub y: f64,
    pub vx: f64,
    pub vy: f64,
    pub category: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LinkState {
    pub source: NodeId,
    pub target: NodeId,
    pub length: f64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SimulationState {
    pub nodes: Vec<NodeState>,
    pub links: Vec<LinkState>,
    pub tick_count: usize,
    pub alpha: f64,
}

pub struct PhysicsEngine {
    pub nodes: Vec<Node>,
    pub links: Vec<Link>,
    pub options: PhysicsOptions,
    pub forces: HashMap<String, ForceFunction>,
    pub tick_count: usize,
}

#[derive(Clone, Debug)]
pub struct PhysicsOptions {
    pub alpha: f64,
    pub alpha_min: f64,
    pub alpha_decay: f64,
    pub alpha_target: f64,
    pub velocity_decay: f64,
    pub width: f64,
    pub height: f64,
}

impl Default for PhysicsOptions {
    fn default() -> Self {
        PhysicsOptions {
            alpha: 1.0,
            alpha_min: 0.001,
            alpha_decay: 0.0228,
            alpha_target: 0.0,
            velocity_decay: 0.4,
            width: 1000.0,
            height: 1000.0,
        }
    }
}

impl PhysicsEngine {
    pub fn new(nodes: Vec<Node>, links: Vec<Link>, options: Option<PhysicsOptions>) -> Self {
        let mut rng = rand::thread_rng();
        let options = options.unwrap_or_default();
        
        // Initialize random positions if not set
        let nodes = nodes
            .into_iter()
            .map(|mut node| {
                if node.x == 0.0 {
                    node.x = rng.gen::<f64>() * options.width;
                }
                if node.y == 0.0 {
                    node.y = rng.gen::<f64>() * options.height;
                }
                
                node.vx = 0.0;
                node.vy = 0.0;
                
                node
            })
            .collect();

        PhysicsEngine {
            nodes,
            links,
            options,
            forces: HashMap::new(),
            tick_count: 0,
        }
    }
    
    pub fn find_node_index(&self, id: &str) -> Option<usize> {
        self.nodes.iter().position(|node| node.id == id)
    }
    
    pub fn add_force(&mut self, name: &str, force_fn: ForceFunction) {
        self.forces.insert(name.to_string(), force_fn);
    }
    
    pub fn tick(&mut self) -> bool {
        // Skip if simulation has cooled down
        if self.options.alpha < self.options.alpha_min {
            return false;
        }
        
        self.tick_count += 1;
        
        // Apply forces to calculate acceleration
        let alpha = self.options.alpha;
        
        // Collect force functions to avoid borrowing conflict
        let force_fns: Vec<ForceFunction> = self.forces.values().copied().collect();
        
        // Apply each force
        for force_fn in force_fns {
            force_fn(self, alpha);
        }
        
        // Update positions using Velocity Verlet integration
        for node in &mut self.nodes {
            if let Some(fx) = node.fx {
                node.x = fx;
                node.vx = 0.0;
            } else {
                node.vx *= self.options.velocity_decay;
                node.x += node.vx;
            }
            
            if let Some(fy) = node.fy {
                node.y = fy;
                node.vy = 0.0;
            } else {
                node.vy *= self.options.velocity_decay;
                node.y += node.vy;
            }
        }
        
        // Cool down simulation
        self.options.alpha += (self.options.alpha_target - self.options.alpha) * self.options.alpha_decay;
        
        true
    }
    
    pub fn initialize_standard_forces(&mut self) {
        self.add_force("charge", Self::many_body_force);
        self.add_force("link", Self::link_force);
        self.add_force("center", Self::center_force);
        self.add_force("collision", Self::collision_force);
    }
    
    pub fn many_body_force(engine: &mut PhysicsEngine, alpha: f64) {
        let strength = -30.0;
        let node_count = engine.nodes.len();
        
        for i in 0..node_count {
            let (charge_i, x_i, y_i) = {
                let node = &engine.nodes[i];
                (node.physical_properties.charge, node.x, node.y)
            };
            
            for j in (i+1)..node_count {
                let (charge_j, x_j, y_j) = {
                    let node = &engine.nodes[j];
                    (node.physical_properties.charge, node.x, node.y)
                };
                
                // Calculate distance vector
                let dx = x_j - x_i;
                let dy = y_j - y_i;
                let distance_squared = dx * dx + dy * dy;
                
                // Skip if nodes are at the same position
                if distance_squared == 0.0 {
                    continue;
                }
                
                // Calculate repulsive force (inverse square law)
                let distance = distance_squared.sqrt();
                let force = strength * charge_i * charge_j / distance_squared;
                
                // Apply force to both nodes (Newton's third law)
                let unit_x = dx / distance;
                let unit_y = dy / distance;
                
                // Apply forces
                let force_x = unit_x * force * alpha;
                let force_y = unit_y * force * alpha;
                
                // Update velocities for both nodes
                engine.nodes[i].vx -= force_x;
                engine.nodes[i].vy -= force_y;
                engine.nodes[j].vx += force_x;
                engine.nodes[j].vy += force_y;
            }
        }
    }
    
    pub fn link_force(engine: &mut PhysicsEngine, alpha: f64) {
        for link in &engine.links.clone() {
            if let (Some(source_idx), Some(target_idx)) = (
                engine.find_node_index(&link.source),
                engine.find_node_index(&link.target),
            ) {
                let (x1, y1, x2, y2) = {
                    let source = &engine.nodes[source_idx];
                    let target = &engine.nodes[target_idx];
                    (source.x, source.y, target.x, target.y)
                };
                
                // Calculate distance vector
                let dx = x2 - x1;
                let dy = y2 - y1;
                let distance = (dx * dx + dy * dy).sqrt();
                
                // Skip if nodes are at the same position
                if distance == 0.0 {
                    continue;
                }
                
                // Calculate spring force (Hooke's law)
                let natural_length = link.physical_properties.length;
                let stiffness = link.physical_properties.stiffness;
                let displacement = distance - natural_length;
                let spring_force = stiffness * displacement;
                
                // Apply force proportional to displacement
                let unit_x = dx / distance;
                let unit_y = dy / distance;
                
                let fx = spring_force * unit_x * alpha;
                let fy = spring_force * unit_y * alpha;
                
                // Apply forces to source and target
                engine.nodes[source_idx].vx += fx;
                engine.nodes[source_idx].vy += fy;
                engine.nodes[target_idx].vx -= fx;
                engine.nodes[target_idx].vy -= fy;
            }
        }
    }
    
    pub fn center_force(engine: &mut PhysicsEngine, alpha: f64) {
        let center_x = engine.options.width / 2.0;
        let center_y = engine.options.height / 2.0;
        let strength = 0.1;
        
        for node in &mut engine.nodes {
            node.vx += (center_x - node.x) * strength * alpha;
            node.vy += (center_y - node.y) * strength * alpha;
        }
    }
    
    pub fn collision_force(engine: &mut PhysicsEngine, alpha: f64) {
        let node_count = engine.nodes.len();
        
        for i in 0..node_count {
            let (radius_i, x_i, y_i) = {
                let node = &engine.nodes[i];
                (node.physical_properties.radius, node.x, node.y)
            };
            
            for j in (i+1)..node_count {
                let (radius_j, x_j, y_j) = {
                    let node = &engine.nodes[j];
                    (node.physical_properties.radius, node.x, node.y)
                };
                
                // Calculate distance vector
                let dx = x_j - x_i;
                let dy = y_j - y_i;
                let distance = (dx * dx + dy * dy).sqrt();
                
                // Skip if nodes are far apart
                let min_distance = radius_i + radius_j;
                if distance >= min_distance {
                    continue;
                }
                
                // Calculate collision response
                let unit_x = dx / distance;
                let unit_y = dy / distance;
                
                // Move nodes apart
                let separation = min_distance - distance;
                let move_x = unit_x * separation * 0.5;
                let move_y = unit_y * separation * 0.5;
                
                // Apply collision forces
                engine.nodes[i].vx -= move_x * alpha;
                engine.nodes[i].vy -= move_y * alpha;
                engine.nodes[j].vx += move_x * alpha;
                engine.nodes[j].vy += move_y * alpha;
            }
        }
    }
    
    pub fn add_category_group_force(&mut self) {
        self.add_force("category_group", Self::category_group_force);
    }
    
    pub fn category_group_force(engine: &mut PhysicsEngine, alpha: f64) {
        // Group nodes by pattern category
        let width = engine.options.width;
        let height = engine.options.height;
        
        let category_groups = HashMap::from([
            ("structural".to_string(), (width * 0.25, height * 0.25)),
            ("process".to_string(), (width * 0.75, height * 0.25)),
            ("relationship".to_string(), (width * 0.5, height * 0.75)),
        ]);
        
        for node in &mut engine.nodes {
            if let Some(&(target_x, target_y)) = category_groups.get(&node.category) {
                node.vx += (target_x - node.x) * 0.01 * alpha;
                node.vy += (target_y - node.y) * 0.01 * alpha;
            }
        }
    }
    
    pub fn run_simulation(&mut self, steps: usize) -> (usize, f64) {
        let mut completed_steps = 0;
        
        for _ in 0..steps {
            if !self.tick() {
                break;
            }
            completed_steps += 1;
        }
        
        (completed_steps, self.options.alpha)
    }
    
    pub fn get_state(&self) -> SimulationState {
        let nodes = self.nodes
            .iter()
            .map(|node| NodeState {
                id: node.id.clone(),
                x: node.x,
                y: node.y,
                vx: node.vx,
                vy: node.vy,
                category: node.category.clone(),
            })
            .collect();
        
        let links = self.links
            .iter()
            .map(|link| LinkState {
                source: link.source.clone(),
                target: link.target.clone(),
                length: link.physical_properties.length,
            })
            .collect();
        
        SimulationState {
            nodes,
            links,
            tick_count: self.tick_count,
            alpha: self.options.alpha,
        }
    }
}